## Importacion de Librerias

In [49]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os

Importa pandas y numpy para manejo de datos.  
Importa matplotlib y seaborn para gráficos.  
Importa os para manejar rutas de archivos.

## Importacion de Datos

In [50]:
#Carpeta de data
DATA_DIR = os.path.join('.', 'data')

Define la carpeta data como ruta base.  
Evita repetir la ruta en cada lectura de CSV.

In [51]:
races = pd.read_csv(os.path.join(DATA_DIR, 'races.csv'))
results = pd.read_csv(os.path.join(DATA_DIR, 'results.csv'))
drivers = pd.read_csv(os.path.join(DATA_DIR, 'drivers.csv'))
constructors = pd.read_csv(os.path.join(DATA_DIR, 'constructors.csv'))
status = pd.read_csv(os.path.join(DATA_DIR, 'status.csv'))
pit_stops = pd.read_csv(os.path.join(DATA_DIR, 'pit_stops.csv'))

Carga las seis tablas del dataset en memoria.  
Cada tabla queda en su propio DataFrame.

## Analisis de tablas

Revisa cada tabla antes de unirla.  
Controla filas, columnas, tipos y nulos.

In [52]:
races.info()

<class 'pandas.DataFrame'>
RangeIndex: 1125 entries, 0 to 1124
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype
---  ------       --------------  -----
 0   raceId       1125 non-null   int64
 1   year         1125 non-null   int64
 2   round        1125 non-null   int64
 3   circuitId    1125 non-null   int64
 4   name         1125 non-null   str  
 5   date         1125 non-null   str  
 6   time         1125 non-null   str  
 7   url          1125 non-null   str  
 8   fp1_date     1125 non-null   str  
 9   fp1_time     1125 non-null   str  
 10  fp2_date     1125 non-null   str  
 11  fp2_time     1125 non-null   str  
 12  fp3_date     1125 non-null   str  
 13  fp3_time     1125 non-null   str  
 14  quali_date   1125 non-null   str  
 15  quali_time   1125 non-null   str  
 16  sprint_date  1125 non-null   str  
 17  sprint_time  1125 non-null   str  
dtypes: int64(4), str(14)
memory usage: 158.3 KB


In [53]:
races.head()

,raceId,year,round,circuitId,name,date,time,url,fp1_date,fp1_time,fp2_date,fp2_time,fp3_date,fp3_time,quali_date,quali_time,sprint_date,sprint_time
0,1,2009,1,1,Australian Grand Prix,2009-03-29,06:00:00,http://en.wikipedia.org/wiki/2009_Australian_G...,\N,\N,\N,\N,\N,\N,\N,\N,\N,\N
1,2,2009,2,2,Malaysian Grand Prix,2009-04-05,09:00:00,http://en.wikipedia.org/wiki/2009_Malaysian_Gr...,\N,\N,\N,\N,\N,\N,\N,\N,\N,\N
2,3,2009,3,17,Chinese Grand Prix,2009-04-19,07:00:00,http://en.wikipedia.org/wiki/2009_Chinese_Gran...,\N,\N,\N,\N,\N,\N,\N,\N,\N,\N
3,4,2009,4,3,Bahrain Grand Prix,2009-04-26,12:00:00,http://en.wikipedia.org/wiki/2009_Bahrain_Gran...,\N,\N,\N,\N,\N,\N,\N,\N,\N,\N
4,5,2009,5,4,Spanish Grand Prix,2009-05-10,12:00:00,http://en.wikipedia.org/wiki/2009_Spanish_Gran...,\N,\N,\N,\N,\N,\N,\N,\N,\N,\N


races tiene 1125 filas y 18 columnas.  
Las columnas de práctica libre quedan como "\N".

* De races interesan 'raceId', 'round','circuitId', 'year', 'name', 'date'

In [54]:
results.info()

<class 'pandas.DataFrame'>
RangeIndex: 26759 entries, 0 to 26758
Data columns (total 18 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   resultId         26759 non-null  int64  
 1   raceId           26759 non-null  int64  
 2   driverId         26759 non-null  int64  
 3   constructorId    26759 non-null  int64  
 4   number           26759 non-null  str    
 5   grid             26759 non-null  int64  
 6   position         26759 non-null  str    
 7   positionText     26759 non-null  str    
 8   positionOrder    26759 non-null  int64  
 9   points           26759 non-null  float64
 10  laps             26759 non-null  int64  
 11  time             26759 non-null  str    
 12  milliseconds     26759 non-null  str    
 13  fastestLap       26759 non-null  str    
 14  rank             26759 non-null  str    
 15  fastestLapTime   26759 non-null  str    
 16  fastestLapSpeed  26759 non-null  str    
 17  statusId         26759 

In [55]:
results.head()

,resultId,raceId,driverId,constructorId,number,grid,position,positionText,positionOrder,points,laps,time,milliseconds,fastestLap,rank,fastestLapTime,fastestLapSpeed,statusId
0,1,18,1,1,22,1,1,1,1,10.0,58,1:34:50.616,5690616,39,2,1:27.452,218.300,1
1,2,18,2,2,3,5,2,2,2,8.0,58,+5.478,5696094,41,3,1:27.739,217.586,1
2,3,18,3,3,7,7,3,3,3,6.0,58,+8.163,5698779,41,5,1:28.090,216.719,1
3,4,18,4,4,5,11,4,4,4,5.0,58,+17.181,5707797,58,7,1:28.603,215.464,1
4,5,18,5,1,23,3,5,5,5,4.0,58,+18.014,5708630,43,1,1:27.418,218.385,1


results tiene 26759 filas y 18 columnas.  
Es la tabla base: un registro por piloto y carrera.

* De results interesan todos, es el dataframe que uso de base para los merges

In [56]:
drivers.info()

<class 'pandas.DataFrame'>
RangeIndex: 861 entries, 0 to 860
Data columns (total 9 columns):
 #   Column       Non-Null Count  Dtype
---  ------       --------------  -----
 0   driverId     861 non-null    int64
 1   driverRef    861 non-null    str  
 2   number       861 non-null    str  
 3   code         861 non-null    str  
 4   forename     861 non-null    str  
 5   surname      861 non-null    str  
 6   dob          861 non-null    str  
 7   nationality  861 non-null    str  
 8   url          861 non-null    str  
dtypes: int64(1), str(8)
memory usage: 60.7 KB


In [57]:
drivers.head()

,driverId,driverRef,number,code,forename,surname,dob,nationality,url
0,1,hamilton,44,HAM,Lewis,Hamilton,1985-01-07,British,http://en.wikipedia.org/wiki/Lewis_Hamilton
1,2,heidfeld,\N,HEI,Nick,Heidfeld,1977-05-10,German,http://en.wikipedia.org/wiki/Nick_Heidfeld
2,3,rosberg,6,ROS,Nico,Rosberg,1985-06-27,German,http://en.wikipedia.org/wiki/Nico_Rosberg
3,4,alonso,14,ALO,Fernando,Alonso,1981-07-29,Spanish,http://en.wikipedia.org/wiki/Fernando_Alonso
4,5,kovalainen,\N,KOV,Heikki,Kovalainen,1981-10-19,Finnish,http://en.wikipedia.org/wiki/Heikki_Kovalainen


drivers tiene 861 pilotos con datos personales.  
Aporta nombre, apellido y nacionalidad.

* De drivers me quedo con 'driverId','forename','surname','nationality'

In [58]:
constructors.info()

<class 'pandas.DataFrame'>
RangeIndex: 212 entries, 0 to 211
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   constructorId   212 non-null    int64
 1   constructorRef  212 non-null    str  
 2   name            212 non-null    str  
 3   nationality     212 non-null    str  
 4   url             212 non-null    str  
dtypes: int64(1), str(4)
memory usage: 8.4 KB


In [59]:
constructors.head()

,constructorId,constructorRef,name,nationality,url
0,1,mclaren,McLaren,British,http://en.wikipedia.org/wiki/McLaren
1,2,bmw_sauber,BMW Sauber,German,http://en.wikipedia.org/wiki/BMW_Sauber
2,3,williams,Williams,British,http://en.wikipedia.org/wiki/Williams_Grand_Pr...
3,4,renault,Renault,French,http://en.wikipedia.org/wiki/Renault_in_Formul...
4,5,toro_rosso,Toro Rosso,Italian,http://en.wikipedia.org/wiki/Scuderia_Toro_Rosso


constructors tiene 212 equipos registrados.  
Aporta nombre y nacionalidad del equipo.

* De constructors interesan 'constructorId', 'name', 'nationality'

In [60]:
status.info()

<class 'pandas.DataFrame'>
RangeIndex: 139 entries, 0 to 138
Data columns (total 2 columns):
 #   Column    Non-Null Count  Dtype
---  ------    --------------  -----
 0   statusId  139 non-null    int64
 1   status    139 non-null    str  
dtypes: int64(1), str(1)
memory usage: 2.3 KB


In [61]:
status.head()

,statusId,status
0,1,Finished
1,2,Disqualified
2,3,Accident
3,4,Collision
4,5,Engine


status traduce el código numérico al motivo de finalización.  
Solo 139 códigos distintos en total.

* De status todas las columnas, 'statusId', 'status'

In [62]:
pit_stops.info()

<class 'pandas.DataFrame'>
RangeIndex: 11371 entries, 0 to 11370
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype
---  ------        --------------  -----
 0   raceId        11371 non-null  int64
 1   driverId      11371 non-null  int64
 2   stop          11371 non-null  int64
 3   lap           11371 non-null  int64
 4   time          11371 non-null  str  
 5   duration      11371 non-null  str  
 6   milliseconds  11371 non-null  int64
dtypes: int64(5), str(2)
memory usage: 622.0 KB


In [63]:
pit_stops.head()

,raceId,driverId,stop,lap,time,duration,milliseconds
0,841,153,1,1,17:05:23,26.898,26898
1,841,30,1,1,17:05:52,25.021,25021
2,841,17,1,11,17:20:48,23.426,23426
3,841,4,1,12,17:22:34,23.251,23251
4,841,13,1,13,17:24:10,23.842,23842


pit_stops registra cada parada individual, no agregada.  
Un piloto puede tener varias filas por carrera.

## Formar dataframe con datos que interesan

In [64]:
races_sub = races[['raceId', 'year', 'round', 'circuitId', 'name', 'date']]
races_sub = races_sub.rename(columns={'name': 'race_name'})

df = results.merge(races_sub, on='raceId', how='left')

Selecciona columnas útiles de races.  
Renombra name a race_name para evitar choques.  
Une results con races por raceId.

In [65]:
drivers_sub = drivers[['driverId', 'forename', 'surname', 'nationality']]
drivers_sub = drivers_sub.rename(columns={'nationality': 'driver_nationality'})

df = df.merge(drivers_sub, on='driverId', how='left')

Selecciona nombre y nacionalidad del piloto.  
Renombra nationality a driver_nationality.  
Une el resultado con drivers por driverId.

In [66]:
constructors_sub = constructors[['constructorId', 'name', 'nationality']]
constructors_sub = constructors_sub.rename(columns={
    'name': 'constructor_name',
    'nationality': 'constructor_nationality'
})

df = df.merge(constructors_sub, on='constructorId', how='left')

Selecciona nombre y nacionalidad del equipo.  
Renombra ambas columnas para distinguirlas del piloto.  
Une el resultado con constructors por constructorId.

In [67]:
df = df.merge(status, on='statusId', how='left')

Une el resultado con status por statusId.  
Traduce el código numérico a texto legible.

In [68]:
pit_stops['duration'] = pd.to_numeric(pit_stops['duration'], errors='coerce')

pit_summary = pit_stops.groupby(['raceId', 'driverId']).agg(
    num_stops=('stop', 'count'),
    total_pit_time=('duration', 'sum')
).reset_index()

df = df.merge(pit_summary, on=['raceId', 'driverId'], how='left')

Convierte duration a tipo numérico.  
Agrupa pit_stops por carrera y piloto.  
Calcula cantidad de paradas y tiempo total en boxes.  
Une el resumen al DataFrame principal.

In [69]:
df.head()

,resultId,raceId,driverId,constructorId,number,grid,position,positionText,positionOrder,points,...,race_name,date,forename,surname,driver_nationality,constructor_name,constructor_nationality,status,num_stops,total_pit_time
0,1,18,1,1,22,1,1,1,1,10.0,...,Australian Grand Prix,2008-03-16,Lewis,Hamilton,British,McLaren,British,Finished,NaN,NaN
1,2,18,2,2,3,5,2,2,2,8.0,...,Australian Grand Prix,2008-03-16,Nick,Heidfeld,German,BMW Sauber,German,Finished,NaN,NaN
2,3,18,3,3,7,7,3,3,3,6.0,...,Australian Grand Prix,2008-03-16,Nico,Rosberg,German,Williams,British,Finished,NaN,NaN
3,4,18,4,4,5,11,4,4,4,5.0,...,Australian Grand Prix,2008-03-16,Fernando,Alonso,Spanish,Renault,French,Finished,NaN,NaN
4,5,18,5,1,23,3,5,5,5,4.0,...,Australian Grand Prix,2008-03-16,Heikki,Kovalainen,Finnish,McLaren,British,Finished,NaN,NaN


Muestra las primeras filas del DataFrame unido.  
Confirma que las uniones funcionaron correctamente.  
num_stops y total_pit_time quedan nulos antes de 2011.

## Limpieza de Datos

Calcula el porcentaje de nulos por columna.  
Ordena las columnas de mayor a menor.

In [70]:
nulos = df.isnull().mean().sort_values(ascending=False) * 100
nulos[nulos > 0]

total_pit_time    79.165888
num_stops         79.165888
dtype: float64

results.csv marca valores faltantes con "\N".  
Convierte esas columnas numéricas a tipo numérico.  
Los valores no convertibles quedan como NaN.

In [71]:
cols_numericas = ['position', 'milliseconds', 'fastestLap', 'rank', 'fastestLapSpeed']

for col in cols_numericas:
    df[col] = pd.to_numeric(df[col], errors='coerce')

pit_stops no registra carreras previas a 2011.  
Marca esas filas antes de rellenar con una columna aparte.  
Rellena num_stops y total_pit_time con cero.

In [72]:
df['tiene_datos_pit'] = df['num_stops'].notna()

df['num_stops'] = df['num_stops'].fillna(0)
df['total_pit_time'] = df['total_pit_time'].fillna(0)

Verifica duplicados usando resultId.  
resultId es la clave única de la tabla results.

In [73]:
duplicados = df.duplicated(subset='resultId').sum()
print(f'Filas duplicadas: {duplicados}')

Filas duplicadas: 0


Convierte date a tipo fecha real.  
Revisa el tipo de dato final de cada columna.

In [74]:
df['date'] = pd.to_datetime(df['date'], errors='coerce')
df.dtypes

resultId                            int64
raceId                              int64
driverId                            int64
constructorId                       int64
number                                str
grid                                int64
position                          float64
positionText                          str
positionOrder                       int64
points                            float64
laps                                int64
time                                  str
milliseconds                      float64
fastestLap                        float64
rank                              float64
fastestLapTime                        str
fastestLapSpeed                   float64
statusId                            int64
year                                int64
round                               int64
circuitId                           int64
race_name                             str
date                       datetime64[us]
forename                          

Revisa la distribución de duration en pit_stops.  
Detecta boxes menores a 10 segundos o mayores a 5 minutos.  
Esos valores suelen ser errores de registro.

In [75]:
pit_stops['duration'].describe()

outliers = pit_stops[(pit_stops['duration'] < 10) | (pit_stops['duration'] > 300)]
outliers

,raceId,driverId,stop,lap,time,duration,milliseconds
